### PR TITLE
[v7] Templates deprecation warning

### PR DESCRIPTION
**Pull request by HaydenCognite**
## Description
Question for those reviewing:
1. Is the message clear enough? I notice inconsistencies with terminology in Cognite and in docs like "Flexible Data Modeling" (FDM), "Data Modeling" (DM), etc. I notice in the docs it is Data Modeling. Is there a consensus I am unaware of?
2. I have seen historically that a migration guide is normally provided between different major versions, however discussing migration from Templates to FDM might be a large discussion. Would a suitable alternative be to provide a link like the one for the docs [here](https://docs.cognite.com/cdf/data_modeling/guides/intro_to_dm/)?

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ Changes are grouped as follows
 - `Security` in case of vulnerabilities.
 
 ## [7.0.0] - 2023-10-24
+### Removed
+- Removed 'function_external_id' argument for creating a Cognite Function Schedule.
```
Comment (haakonvt) [inline CHANGELOG.md:22]:
```suggestion
- Removed `function_external_id` argument for creating a Cognite Function schedule (only worked with API keys.)
```

File: MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -8,6 +8,12 @@ Changes are grouped as follows:
 - `Changed` for changes that do not fall into any other category
 - `Optional` for new, optional methods/features that you should be aware of - *and could take advantage of*
 
+## From v6 to v7
+
+### Removed
+
+- Removed the deprecated 'function_external_id' parameter when creating a Cognite Function Schedule. Future instantiations are to be done without this parameter now.
```
Comment (haakonvt) [inline MIGRATION_GUIDE.md:15]:
```suggestion
- Removed the deprecated `function_external_id` parameter when creating a Cognite Function schedule. This only worked when using API keys - which are long gone.
```

File: cognite/client/_api/assets.py [hunk 3]
```diff
@@ -602,7 +602,7 @@ def create_hierarchy(
         Args:
             assets (Sequence[Asset] | AssetHierarchy): List of assets to create or an instance of AssetHierarchy.
             upsert (bool): If used, already existing assets will be updated instead of an exception being raised. You may control how updates are applied with the 'upsert_mode' argument.
-            upsert_mode (Literal["patch", "replace"]): Only applicable with upsert. Pass 'patch' to only update fields with non-null values (default), or 'replace' to do full updates (unset fields become null or empty).
+            upsert_mode (Literal['patch', 'replace']): Only applicable with upsert. Pass 'patch' to only update fields with non-null values (default), or 'replace' to do full updates (unset fields become null or empty).
```
Comment (haakonvt) [inline cognite/client/_api/assets.py:605]: The linter accepts these single quote changes? 🤔

File: cognite/client/_api/functions.py [hunk 4]
```diff
@@ -63,11 +63,11 @@ def _get_function_internal_id(cognite_client: CogniteClient, identifier: Identif
...
 
 
-def _get_function_identifier(function_id: int | None, function_external_id: str | None) -> Identifier:
-    identifier = IdentifierSequence.load(function_id, function_external_id, id_name="function")
+def _get_function_identifier(function_id: int | None) -> Identifier:
+    identifier = IdentifierSequence.load(function_id, id_name="function")
     if identifier.is_singleton():
         return identifier[0]
-    raise AssertionError("Exactly one of function_id and function_external_id must be specified")
+    raise AssertionError("Argument function_id must be specified")
```
Comment (haakonvt) [inline cognite/client/_api/functions.py:70]: Functions still can be referenced with external id, so these changes must be reverted. It is only attaching a schedule that requires "id only".

File: cognite/client/_api/functions.py [hunk 5]
```diff
@@ -1041,7 +1024,7 @@ def create(
                 ... )
 
         """
-        _get_function_identifier(function_id, function_external_id)
+        _get_function_identifier(function_id)
```
Comment (haakonvt) [inline cognite/client/_api/functions.py:1027]: This should no longer use that function, as the error message that can be raised would be misleading:
`"Exactly one of function_id and function_external_id must be specified"`

File: cognite/client/_api/templates.py [hunk 6]
```diff
@@ -26,6 +27,13 @@
     from cognite.client.config import ClientConfig
 
 
+def templates_deprecation_warning() -> None:
+    return warnings.warn(
```
Comment (haakonvt) [inline cognite/client/_api/templates.py:31]:
```suggestion
def _templates_deprecation_warning() -> None:
    warnings.warn(
```

File: cognite/client/_api/templates.py [hunk 7]
```diff
@@ -26,6 +27,13 @@
...
+        "Templates will be removed in future versions of the SDK. Please migrate to Flexible Data Modeling.",
```
Comment (haakonvt) [inline cognite/client/_api/templates.py:32]: This sounds better to me as a norwenglishman 😂  
```suggestion
        "Templates will be removed in a future version of the SDK. Please migrate to Data Modeling. "
        "Read more at: https://docs.cognite.com/cdf/data_modeling/",
```

**Review by haakonvt — CHANGES_REQUESTED** [review]
PR title only mentions templates, consider splitting into 2 PRs or updating. Please check that the linter is happy with all these single-quote changes.

**Comment by haakonvt** [comment]
Fyi: @HaydenCognite I rebased your branch on `release-v7` so that CI checks run.

File: cognite/client/_api/templates.py (hunk 7 above)
Comment (HaydenCognite) [reply cognite/client/_api/templates.py:32]: I actually had written that suggestion initially, but I decided to copy what was written in some other deprecation messages elsewhere. Who knows, maybe Templates come back (not likely, but y'know what I mean).

File: cognite/client/_api/assets.py (hunk 3 above)
Comment (HaydenCognite) [reply cognite/client/_api/assets.py:605]: I ran the pre-commit on everything. 🤷‍♂️ It must have changed them.

File: cognite/client/_api/functions.py (hunk 4 above)
Comment (HaydenCognite) [reply cognite/client/_api/functions.py:70]: Ah, my misunderstanding. I will fix these later today.

**Review by haakonvt — APPROVED** [review]

**Pull request merged**